ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/helpers/BeaconStateMutators.java

```diff
@@ -98,18 +110,85 @@ public void initiateValidatorExit(MutableBeaconState state, int index) {
         .set(
             index,
             validator
-                .withExitEpoch(exitQueueEpoch)
+                .withExitEpoch(validatorExitContext.exitQueueEpoch)
                 .withWithdrawableEpoch(
-                    exitQueueEpoch.plus(specConfig.getMinValidatorWithdrawabilityDelay())));
+                    validatorExitContext.exitQueueEpoch.plus(
+                        specConfig.getMinValidatorWithdrawabilityDelay())));
+  }
+
+  /**
+   * This function implements an optimized version of exitQueueEpoch and exitQueueChurn calculation,
+   * compared to the `initiateValidatorExit` reference implementation.
+   *
+   * @param state
+   * @return
+   */
+  public Supplier<ValidatorExitContext> createValidatorExitContextSupplier(
+      final BeaconState state) {
+    return Suppliers.memoize(
+        () -> {
+          final ValidatorExitContext validatorExitContext =
+              new ValidatorExitContext(beaconStateAccessors.getValidatorChurnLimit(state));
+
+          validatorExitContext.exitQueueEpoch = UInt64.ZERO;
+
+          final List<Validator> exitedValidatorsInExitQueueEpoch = new ArrayList<>();
+
+          for (Validator validator : state.getValidators()) {
+            final UInt64 validatorExitEpoch = validator.getExitEpoch();
+            if (validatorExitEpoch.equals(FAR_FUTURE_EPOCH)) {
+              continue;
+            }
+            if (validatorExitEpoch.isGreaterThan(validatorExitContext.exitQueueEpoch)) {
+              exitedValidatorsInExitQueueEpoch.clear();
+              validatorExitContext.exitQueueEpoch = validatorExitEpoch;
+              exitedValidatorsInExitQueueEpoch.add(validator);
+              continue;
+            }
+            if (validatorExitEpoch.equals(validatorExitContext.exitQueueEpoch)) {
+              exitedValidatorsInExitQueueEpoch.add(validator);
+            }
+          }
+          final UInt64 activationExitEpoch =
+              miscHelpers.computeActivationExitEpoch(beaconStateAccessors.getCurrentEpoch(state));
+
+          if (activationExitEpoch.isGreaterThan(validatorExitContext.exitQueueEpoch)) {
+            validatorExitContext.exitQueueEpoch = activationExitEpoch;
+            validatorExitContext.exitQueueChurn = UInt64.ZERO;
+          } else {
+            validatorExitContext.exitQueueChurn =
+                UInt64.valueOf(exitedValidatorsInExitQueueEpoch.size());
+          }
+
+          return validatorExitContext;
+        });
+  }
+
+  public static class ValidatorExitContext {
```

Comment:
I may sound little boring, but I don't like direct value access both for writing and reading here. Not sure this should be improved, just some eye catching for me

Reply:
I tend to be not super strict on that in cases like this where the fields are private, I see them like private fields in the outer class actually.